### PR TITLE
wip: snapshot testing (fixes #24)

### LINKED ISSUE #24: add snapshot testing
@aladdin-add Is this something you can do? Just add the basic setup and I can port / fix all test files? For example to similar to what you see in the Tenko repo. .md files.

### PR DESCRIPTION
**Pull request by aladdin-add**
TODO:

- [ ] remove unused options
- [ ] support loading src/
- [ ] support recovery mode

### CODE REVIEW

**Comment by KFlash** [comment]
Lots of code! Did you get it working?

**Comment by aladdin-add** [comment]
not yet! just copied it from tenko repo.

it seems very complicated, I prefer putting it another package. thoughts?

**Comment by KFlash** [comment]
Look at the same others preval library. It is using same but simplified code. No need to create a package.
Current code inckude a lot to be deleted. Like acorn and Babel testing

**Comment by KFlash** [comment]
@aladdin-add any update on this?

**Comment by aladdin-add** [comment]
I just make it work with valid cases. ╮(￣▽￣)╭

**Comment by aladdin-add** [comment]
https://github.com/escaya/escaya/blob/4f165af1bae0f13ed8cfc1fba61321eb94b6de04/tests/testcases/var/binding_generic/as_a_statement/regular_vars/var_var_with_init_asi.md

just the same format as tenko. is it something you imagined?

**Comment by KFlash** [comment]
Looks nice. Huge, but looks nice. How do you auto generate a file and write a test file? And refresh?

File: package.json [hunk 1]
```diff
@@ -46,7 +46,9 @@
     "test:watch:verbose": "npm run test:watch -- -R spec",
     "prepare-nightly": "node scripts/bump-dev-version",
     "coverage": "cross-env TS_NODE_PROJECT=\"test/tsconfig.json\" nyc -n \"src/**/*.ts\" -e .ts -i ts-node/register -r text-summary -r lcov -r html npm test",
-    "post_coverage": "cross-env cat ./coverage/lcov.info | coveralls"
+    "post_coverage": "cross-env cat ./coverage/lcov.info | coveralls",
+    "snapshot": "node tests/run_tests.mjs",
+    "snapshot:update": "node tests/run_tests.mjs -u"
```
Comment (aladdin-add) [inline package.json:51]: this is to update the snaphot :). and I think we need another script to auto gen the cases.

Comment (KFlash) [reply package.json:51]: I see, but how do you write a test and auto gen?

Comment (aladdin-add) [reply package.json:51]: e.g. https://github.com/pvdz/tenko/blob/bffb35fbdc3fb63cd052d2d85a9b8df0bcf31bbe/tests/testcases/group_or_arrow/arrow/spread/arrow_destruct_of_obj/autogen.md

File: tests/run_tests.mjs [hunk 2]
```diff
@@ -0,0 +1,1209 @@
...
+if (INPUT_OVERRIDE) {
+  LOG('Using override input and only testing that...');
+  if (!a && !b && !c) LOG('Sloppy mode implied (override with --strict or --module)');
+  if (b && !a && !c) LOG('Testing in strict mode only');
+  if (c && !a && !b) LOG('Testing in module goal only');
+  if (ENABLE_ANNEXB) LOG('Testing with Annex B syntax extensions enabled');
+  LOG('=============================================\n');
+} else {
+  if (!RUN_VERBOSE_IN_SERIAL) console.time('$$ Test search discovery time');
+  getTestFiles(path.join(dirname, 'testcases'), '', files, true);
```
Comment (KFlash) [inline tests/run_tests.mjs:1198]: try 'path.resolve('tests/testcases')'

**Comment by KFlash** [comment]
I think the challenge will be to get coverage working together with this plus  test "recovery mode". Need another settings for it because it never throws. How would you know if it should throw? Through diagnostics?

**Comment by aladdin-add** [comment]
how did you get it tested now? (I had very little knowledge of "recovery mode".)

File: tests/run_tests.mjs (hunk 2 above)
Comment (aladdin-add) [reply tests/run_tests.mjs:1166]: why using `resolve` here? - it's always existed.

**Comment by KFlash** [comment]
> how did you get it tested now? (I had very little knowledge of "recovery mode".)

I just check if diagnostic array is empty on rootnode, If not empty then it has failed. Can improve this and iterate through the array I guess and throw each diagnostics or update the .md file to recognise expected diagnostics.

I think the last one is the best suggestion - update the .md to recognise expected diagnostics in array instead of throw

**Comment by KFlash** [comment]
@aladdin-add Take a look at Preval. Simplified version of same test runner

**Comment by KFlash** [comment]
@aladdin-add Did you manage to clean up the code etc?

**Comment by aladdin-add** [comment]
preval? you mean https://github.com/kentcdodds/babel-plugin-preval ?

**Comment by KFlash** [comment]
> preval? you mean https://github.com/kentcdodds/babel-plugin-preval ?

Yes

File: tests/run_tests.mjs (hunk 2 above)
Comment (KFlash) [reply tests/run_tests.mjs:1166]: windows 10

**Comment by KFlash** [comment]
any update on this?

**Comment by KFlash** [comment]
@aladdin-add friendly ping

**Comment by KFlash** [comment]
@aladdin-add friendly ping

**Comment by aladdin-add** [comment]
As for preveal, I just had a quick look. seems there are some features missing (e.g. auto-gen), and not well-tested.

the current implementation is ready to use.  you can port all tests to the md files.
if you prefer another way, you can change it later - but it's not blocker to writing tests.

**Comment by KFlash** [comment]
Okay. I just noticed Babel, acorn tests and ECMA versions. Should be removed?

**Review by KFlash — APPROVED** [review]
Seems okay, but loads of extra code that's not needed?

**Comment by KFlash** [comment]
@aladdin-add should try to support code coverage as we do with mocha

**Comment by aladdin-add** [comment]
good to have!👍

**Comment by KFlash** [comment]
@aladdin-add Off topic. Do you have lots of spare time and/ or do you want to contribute on something that is going to be a game changer? I re-invented TypeScript and built a parser in the way TS should have been, but the code aim to replace prettify,eslint, babel and much more.  The parser code is 99.9% done now, but still playing around with prettify and transformers

**Comment by KFlash** [comment]
@aladdin-add Put this on hold for awhile ;)

**Comment by aladdin-add** [comment]
seems this one can be closed, as you had rewritten your own runner.

will port to this repo.

**Pull request closed without merging**